README.md

# Evolutions

The evolutions package provides a facility similar to
[play-evolutions](https://www.playframework.com/documentation/2.7.x/Evolutions)
([source](https://github.com/playframework/playframework/tree/master/persistence/play-jdbc-evolutions/src/main/scala/play/api/db/evolutions))
for use with Python.  It differs from other similar tools in its simple,
lightweight philosophy and its pure SQL orientation.  It enforces a linear
history in schema development with no branching or dependencies.  Migrations
are contained in SQL files and run automatically as needed, with no code
embedding or hand-tuned invocations needed.  Evolutions plays well with source
control as well as differences between development and production
environments.  Finally, it's not tied to any application framework so can be
used anywhere that Python 3 and the target database are available, regardless
of whether the using application is itself implemented in Python.

## Ups

The basic idea is that, as the database schema is developed incrementally, the
changes are placed in *ups* files named `1.sql` (the original schema),
`2.sql`, `3.sql`, etc..  When the tool is invoked (usually just before the app
starts up, see Usage), a table is checked in the database to see which
of these scripts have been run, and any missing are run in sequence.  The
scripts should contain both schema changes and data migration.  This way, the
app and the schema can be developed together, without worrying about manual
database updates, legacy data support, or schema-code synchronization.

### Example

An application is deployed with its schema in a file `1.sql`,
containing a lot of `CREATE TABLE` statements.  Some development is done and
`2.sql` with some `ALTER TABLE` statements is written.  When this is deployed
and the evolutions tool run before application startup, it detects that
`1.sql` has already been run, but `2.sql` has not, so it runs that.  Later on,
`3.sql` and `4.sql` are developed but reach deployment together.  When the
evolutions tool is run, it runs first `3.sql` then `4.sql` in sequence.

## Downs

In addition to the *ups*, a set of corresponding *downs* files `1-downs.sql`,
`2-downs.sql`, `3-downs.sql` are created during development, which undo the
effects of the corresponding ups files.  If the tool detects that a change has
been made to an ups file, then it will rerun it.  However, its former version,
as well as any later ups that have been run "on top" of it, must be undone
first.  So the tool first runs these in reverse, using versions cached in the
database itself in case *those* have changed as well, then runs back up the
sequence starting from the first modified file.  This process allows tweaking
changes during development, without needing to manually adjust the database
each time.  It also enables collaboration, so that, for example, a single
`#.sql` file can be used by everyone for an entire sprint (merging in changes
to a shared git branch).

### Example

The application in the first example continues being developed. In addition to
the ups files mentioned earlier, `1-downs.sql` (a bunch of `DROP TABLE`
statements), `2-downs.sql` (`ALTER TABLE` statements undoing the alterations
in `2.sql`), `3-downs.sql`, and `4-downs.sql` were all written.  At one point,
someone realizes there was a mistake in `3.sql` and it should be fixed.  They
modify this file, as well as `3-downs.sql`, then deploy the application.

When the evolutions tool runs, it detects that `3.sql` has changed.  It then
runs `4-downs.sql` and `3-downs.sql` (in that order) to get to a point where
it can correctly apply the new version of `3.sql`.  Then it runs that script,
and finally `4.sql` to bring the database fully up to date.

## Production

In general, although running downs will ensure a consistent database schema,
some data loss is often unavoidable, simply because elements of the schema are
lost through downgrade, meaning the data cannot be preserved.  **For this
reason, you should not run downs in production. Ever.** Once you deploy an
evolution stage to production, you should freeze it, so evolutions will not
run it again.  In addition, you can and should enable a safety check by adding
the `--prod` switch when invoking (see Usage).  If the tool encounters an ups
script change when this switch is active, it will abort rather than run any
downs, so you can address the situation manually.

In this case, you will end up in a situation where the database is out of sync
with the evolutions scripts.  This can also occur if evolutions is only taken
into use.  This can be remedied by using the `--skip=<indices>` switch to the
tool, which will cause the tool to assume the scripts with the given
comma-separated indices have already been run, and mark them so in the
database, without actually running them.

### Example

The deployment mentioned in the preceding examples as assumed to be on a
development environment, where it's OK to lose data in the case *downs*
scripts need to be run.  However, in the production deployment, the evolutions
tool is invoked with the `--prod` argument.  Hence, in the situation described
in the previous example where `3.sql` has changed, it will abort and return an
error code.  The user will then need to address the situation manually.
Possibly they will apply the new change needed from `3.sql` manually, taking
care to preserve data.

Then they can invoke the tool using `--skip=3`.  This will cause it to mark
`3.sql` has having been "run" in its new modified version, *without actually
running it*.  Then, after noting that `4.sql` has not changed, the database is
now considered to be up to date.  (Subsequent invocations do not need the
`--skip` argument and will not trigger an abort, unless there has been a new,
different change of an already-run script.)


# Database Support

MySQL, PostgreSQL, and Sqlite are supported.  The database is accessed through
a combination of the command line clients and
[DB-API2](https://www.python.org/dev/peps/pep-0249/), using
[mysql-connector-python](https://dev.mysql.com/doc/connector-python/en/),
[PsycoPg2](http://initd.org/psycopg/docs/), and
[sqlite3](https://docs.python.org/3.8/library/sqlite3.html).  (These are not
listed as dependencies of this package, so you should install the one needed
for your own case yourself.)


## Transactions

The evolutions scripts themselves are run by invoking the command line client
for the database being used (e.g., `psql` for Postgres).  Transactionality is
therefore under control of the script itself.

When using MySQL or PostgreSQL (but not Sqlite), the evolutions table is
locked when the script starts, so any parallel instances of the script started
up will wait until the first one completes, and then will find the evolutions
already run and do nothing.


# Install

    pip3 install evolutions

-or- to get development version:

    pip3 install https://github.com/arobertn/evolutions/archive/master.tar.gz


# Usage

The tool is invoked via a Python 3 command line script, and should be called
just before or as part of starting the application, or, in an auto-deploy
environment, whenever the schema files have been changed.

    ./evolutions.py <db_url> <db_user> <db_pass> <evolutions_dir> [--skip=<stages>] [--prod]
        --skip=<stages> = comma-separated indices to assume already run
        --prod          = abort if downs need to be run (for production)

- *db\_url:* e.g.: `mysql://localhost:3306/dbname`,
                 `postgresql://localhost:5432/dbname`,
                 `sqlite:/absolute/path/to/database.db`
- *db\_user, db\_pass:* db\_pass is ignored for Postgres (must use .pgpass
  file), and both db\_user and db\_pass are ignored for Sqlite; pass empty
  strings ("")
- *evolutions_dir:* directory containing the #.sql/#-downs.sql files; can be
  relative or absolute path
- *--skip:* comma-separated list of stage indices to skip running if you have
  a database which has already had one or more of the #.sql files run on it;
  will insert rows to the `evolutions` table to make it up to date, but will
  not actually run the scripts
- *--prod:* if given, tool will abort immediately if it determines any downs would
  need to be run; database is not touched


# Implementation

The evolutions tool operates by collecting the SHA1 hash of each ups and downs
script in the evolutions directory, and storing these values, together with
the script contents themselves, in a dedicated table (named `evolutions`) in
the database.  Decisions on which ups and downs scripts to run are made by
comparing the database record and the scripts found in the directory, and
updates are made according to the runs.


# Development

On [GitHub](https://github.com/arobertn/evolutions).


CHANGELOG.md
# Changelog

## 0.8.3

2022-12-27

- psql improvements: host/post handling, better detection of SQL failures

## 0.8.2

2019-11-26

- Lock the evolutions table (PostgreSQL, MySQL) so concurrent runs
  (e.g. starting a clustered application) should work correctly
- Tests now compatible with MySQL 5.6, 5.7
- Mark package as Python 3 only

## 0.8.1

2019-11-23

- Small packaging improvements

## 0.8.0

2019-11-22

- Initial release


evolutions/test/runtests.py
#!/usr/bin/env python3

# Invoke from top-level dir

import os, os.path as path, subprocess, sys, unittest

sys.path.append('.')


class TestEvolutions_MySQL(unittest.TestCase):

    # This and next are seriously abusing Python's loose OO implementation
    @classmethod
    def setUpClass(cls):
        db_name = 'evtest'
        db_user = 'evtestuser'
        db_pass = 'evtestpass'
        cls.db_cmd = ['./evolutions/evolutions.py',
                      'mysql://localhost:3306/' + db_name,
                      db_user, db_pass]
        cls.db_check_cmd = ['mysql', '-u', db_user,
                            '--password=' + db_pass, db_name,
                            '--skip-column-names']
        print("\nMySQL Tests\n", file=sys.stderr)

    @classmethod
    def tearDownClass(cls):
        cls.do_db_check(cls, "DROP TABLE soup;")
        cls.do_db_check(cls, "DROP TABLE evolutions;")


    def do_db_check(self, query, expected_result=None):
        db_proc = subprocess.Popen(self.db_check_cmd,
                                   stdout=subprocess.PIPE,
                                   stderr=sys.stderr,
                                   stdin=subprocess.PIPE)
        output, _ = db_proc.communicate(query.encode('utf-8'))
        if db_proc.returncode != 0:
            raise Exception('check query "' + query + '" failed: '
                            + output.decode('utf-8'))
        if expected_result:
            self.assertEqual(output.strip().decode('utf-8'), expected_result)

    def setUp(self):
        print("\nTest '%s'\n" % (self._testMethodName), file=sys.stderr)


    # Load a single stage correctly, no-op on rerun
    def test_case_1(self):
        subprocess.check_call(self.db_cmd + ['evolutions/test/case_1'])
        self.do_db_check("SELECT COUNT(*) FROM SOUP;", "2")

        # Repeat should be ok, be a no-op
        subprocess.check_call(self.db_cmd + ['evolutions/test/case_1'])
        self.do_db_check("SELECT COUNT(*) FROM SOUP;", "2")

    # Add a second and third stage to the first
    def test_case_2(self):
        subprocess.check_call(self.db_cmd + ['evolutions/test/case_2'])
        self.do_db_check("SELECT COUNT(*) FROM SOUP;", "4")

    # Change stage 1, but run in prod mode, causing abort
    def test_case_3a(self):
        ret = subprocess.call(self.db_cmd + ['evolutions/test/case_3',
                                             '--prod'])
        self.assertEqual(ret, 1)
        self.do_db_check("SELECT COUNT(*) FROM SOUP;", "4")

    # Change stage 1, triggering downs and rerun of all stages
    def test_case_3b(self):
        subprocess.check_call(self.db_cmd + ['evolutions/test/case_3'])
        self.do_db_check("SELECT COUNT(*) FROM SOUP;", "5")

    # Check file misconfiguration (missing downs)
    def test_case_4(self):
        ret = subprocess.call(self.db_cmd + ['evolutions/test/case_4'])
        self.assertEqual(ret, 1)

    # Check file misconfiguration (incorrect naming)
    def test_case_5(self):
        ret = subprocess.call(self.db_cmd + ['evolutions/test/case_5'])
        self.assertEqual(ret, 1)

    # Check file misconfiguration (skipped stage)
    def test_case_6(self):
        ret = subprocess.call(self.db_cmd + ['evolutions/test/case_6'])
        self.assertEqual(ret, 1)

    # Add stage 5 but skip running it
    def test_case_7(self):
        subprocess.check_call(self.db_cmd + ['evolutions/test/case_7',
                                             '--skip=5'])
        self.do_db_check("SELECT COUNT(*) FROM SOUP;", "5")
        self.do_db_check("SELECT COUNT(*) FROM evolutions;", "5")
        # Skip not needed after that
        subprocess.check_call(self.db_cmd + ['evolutions/test/case_7'])
        self.do_db_check("SELECT COUNT(*) FROM evolutions;", "5")

    # Modify stage 3, but skip running it, in prod mode
    def test_case_8(self):
        subprocess.check_call(self.db_cmd + ['evolutions/test/case_8',
                                             '--skip=3', '--prod'])
        self.do_db_check("SELECT COUNT(*) FROM SOUP;", "5")
        self.do_db_check("SELECT COUNT(*) FROM evolutions;", "5")
        # Skip not needed after that
        subprocess.check_call(self.db_cmd + ['evolutions/test/case_8',
                                             '--prod'])
        self.do_db_check("SELECT COUNT(*) FROM evolutions;", "5")


class TestEvolutions_PostgreSQL(TestEvolutions_MySQL):

    @classmethod
    def setUpClass(cls):
        db_name = 'evtest'
        db_user = 'evtestuser'
        db_pass = ''
        cls.db_cmd = ['./evolutions/evolutions.py',
                      'postgresql://localhost:5432/' + db_name,
                      db_user, db_pass]
        cls.db_check_cmd = ['psql', '-h', 'localhost', '-U', db_user, db_name,
                            '--tuples-only', '--quiet']
        print("\nPostgreSQL Tests\n", file=sys.stderr)


class TestEvolutions_Sqlite(TestEvolutions_MySQL):

    @classmethod
    def setUpClass(cls):
        db_name = 'evtest.db'
        cls.db_cmd = ['./evolutions/evolutions.py',
                      'sqlite:' + path.join(os.getcwd(),
                                            'evolutions', 'test', db_name),
                      '""', '""']
        cls.db_check_cmd = ['sqlite3', path.join('evolutions', 'test', db_name)]
        print("\nSqlite Tests\n", file=sys.stderr)


if __name__ == '__main__':
    unittest.main()


evolutions/test/init_psql_test.sh


createuser -U postgres -d -PEe evtestuser
createdb -U evtestuser -E UTF8 evtest

echo "localhost:5432:evtest:evtestuser:evtestpass" >> ~/.pgpass


evolutions/evolutions.py
#!/usr/bin/env python3

# evolutions: Provides play-evolutions-like functionality for Python
#             https://www.playframework.com/documentation/2.7.x/Evolutions
# usage: ./evolutions.py <db_url> <db_user> <db_pass> <evolutions_dir>

import glob, hashlib, logging, os.path as path, re, subprocess, sys


# Clean, uniform informational and error output
logging.basicConfig(level='INFO', format='evolutions: %(message)s')
logger = logging.getLogger()


# Holds a DB connection and info it was created from
class DBConn:
    def __init__(self, db_type, cmd, host, port, db_name, user, pw, conn,
                 param):
        self.db_type = db_type
        self.cmd     = cmd
        self.host    = host
        self.port    = port
        self.db_name = db_name
        self.user    = user
        self.pw      = pw
        self.conn    = conn
        self.param   = param

    # Ugh, different DB-API2 impls use '?' or '%s' for parameter wildcard
    def fix_params(self, stmt):
        return stmt.replace('_?', self.param)

    # Wraps to handle parameter formatting
    def execute(self, stmt, *args):
        db = self.conn.cursor()
        db.execute(self.fix_params(stmt), *args)
        return db


# Holds info on a single evolution stage (both ups and downs)
class Stage:
    def __init__(self, idx, apply_hash, revert_hash,
                 apply_script, revert_script, applied_at):
        self.idx = idx
        self.apply_hash = apply_hash
        self.revert_hash = revert_hash
        self.apply_script = apply_script
        self.revert_script = revert_script
        self.applied_at = applied_at

    def __str__(self):
        return 'Stage %d (apply len = %d, revert len = %d)' % (
            self.idx, len(self.apply_script), len(self.revert_script))


# Utility
def read_textfile(fname):
    with open(fname, 'r', encoding='utf-8') as f:
        return f.read()


# Return DBConn wrapping DB-API2 conn (https://python.org/dev/peps/pep-0249/)
def get_connection(url, user, pw):
    url_re = re.compile(r'([^:]+)://([^:]+)(:([0-9]+))?/(.+)')
    file_re = re.compile(r'([^:]+):(/.+)')
    match = url_re.match(url)
    if match:
        db_type, host, port, db_name = (match.group(1), match.group(2),
                                        match.group(4), match.group(5))
    else:
        match = file_re.match(url)
        if match:
            db_type, host, port, db_name = (match.group(1), None,
                                            None, match.group(2))
        else:
            raise Exception("Unrecognized DB URL format: '" + url + "'")
    if db_type == 'mysql':
        import mysql.connector
        port = port or '3306'; # String because that's what match would have yielded
        conn = mysql.connector.connect(user=user, password=pw,
                                       host=host, port=port, database=db_name,
                                       charset='utf8', autocommit=False)
        cmd = ['mysql', '-u', user, '--password='+pw, db_name]
        param = '%s'
    elif db_type == 'postgresql':
        import psycopg2, os
        port = port or '5432'; # String because that's what match would have yielded
        conn = psycopg2.connect(user=user, password=pw,
                                host=host, port=port, database=db_name)
        conn.set_session(autocommit=False)
        os.environ['PGPASSWORD'] = pw # Password via env ins
        cmd = ['psql', '-h', host or 'localhost', '-p', port, '-U', user, db_name]
        param = '%s'
    elif db_type == 'sqlite':
        import sqlite3
        # Note: URL is assumed to contain absolute path in this case
        conn = sqlite3.connect(db_name, isolation_level=None)
        cmd = ['sqlite3', db_name]
        param = '?'
    else:
        raise Exception("Unsupported database type: '" + db_type + "'")

    return DBConn(db_type, cmd, host, port, db_name, user, pw, conn, param)


# Connects to database and ensures evolutions table present
def connect_and_ensure(url, user, pw):
    dbConn = get_connection(url, user, pw)
    dbConn.execute('''
      CREATE TABLE IF NOT EXISTS evolutions (
          id             INT NOT NULL PRIMARY KEY,
          applied_at     TIMESTAMP NOT NULL,
          apply_hash     VARCHAR(64) NOT NULL,
          revert_hash    VARCHAR(64) NOT NULL,
          apply_script   TEXT,
          revert_script  TEXT )
    ''')

    # We lock access to evolutions table to prevent concurrent runs of
    # evolutions from interfering. Not supported for sqlite since only
    # full database lock is supported, but for same reason not an issue there.
    if dbConn.db_type == 'mysql':
        dbConn.execute("LOCK TABLES evolutions WRITE")
    elif dbConn.db_type == 'postgresql':
        dbConn.execute("LOCK TABLE evolutions IN ACCESS EXCLUSIVE MODE NOWAIT")
    return dbConn


# Sha1 of a file; use only with small files, else chunk it
def sha1_file(fname):
    with open(fname, 'rb') as f:
        return hashlib.sha1(f.read()).hexdigest()


# Reads in and collects hashes for all evolutions files in a directory
def scan_dir_stages(ev_dir):
    numeric_re = re.compile(r'([0-9]+)(-downs)?\.sql')
    sql_files = glob.glob(path.join(ev_dir, '*.sql'))
    ups = []
    downs = []

    for sql_file in sql_files:
        m = numeric_re.fullmatch(path.basename(sql_file))
        if m:
            idx = int(m.group(1))
            is_ups = m.group(2) is None
            if is_ups:
                ups.append(idx)
            else:
                downs.append(idx)
    ups = sorted(ups)
    downs = sorted(downs)

    if ups != downs:
        raise Exception("Ups and downs SQL files are not in correspondence.")

    stages = []
    for idx in ups:
        ups_f = path.join(ev_dir, str(idx) + '.sql')
        downs_f = path.join(ev_dir, str(idx) + '-downs.sql')
        stages.append(Stage(idx, sha1_file(ups_f), sha1_file(downs_f),
                            read_textfile(ups_f), read_textfile(downs_f), None))
    return stages


def scan_db_stages(dbConn):
    res = dbConn.execute('''
        SELECT id, apply_hash, revert_hash, apply_script, revert_script, applied_at FROM evolutions ORDER BY id ASC;
    ''')

    stages = []
    for row in res.fetchall():
        stages.append(Stage(row[0], row[1], row[2], row[3], row[4], row[5]))

    return stages


# Check if stages start with 1 and go in sequence, assuming sorted
def check_stages(stages, src):
    if stages:
        indices = [ stage.idx for stage in stages ]
        if indices[0] != 1 or indices[-1] != len(indices):
            raise Exception('Illegal stage sequence ' + str(indices)
                            + ' in ' + src)
    return stages


# Invoke db command to execute script (DBAPI has no multistatement support)
def execute_script(idx, script_str, dbConn):
    db_proc = subprocess.Popen(dbConn.cmd,
                               stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE,
                               stdin=subprocess.PIPE)
    output, error = db_proc.communicate(script_str.encode('utf-8'))
    # Note: Errors can happen with the return value from psql utility still being zero.
    # Therefore, explicitly look for errors in collected stderr content.
    if (db_proc.returncode != 0) or (error):
        raise Exception('evolutions: script ' + str(idx) + "\n\t" +
                        output.decode('utf-8').replace("\n", "\n\t") + "\n\t" +
                        error.decode('utf-8').replace("\n", "\n\t")
        )


# Executes given stage downs, and removes stage from db
def run_and_remove_downs(stage, dbConn):
    logger.info('Running downs for stage %d', stage.idx)
    execute_script(stage.idx, stage.revert_script, dbConn)
    dbConn.execute('DELETE FROM evolutions WHERE id = _?', [stage.idx])


# Executes ups (unless in list to skip as already run) and record to db
def run_and_add_ups(stage, skip, dbConn):
    if stage.idx in skip:
        logger.warning('Force skip running ups for stage %d', stage.idx)
    else:
        logger.info('Running ups for stage %d', stage.idx)
        execute_script(stage.idx, stage.apply_script, dbConn)

    dbConn.execute('''
        INSERT INTO evolutions (id, applied_at, apply_hash, revert_hash,
                                apply_script, revert_script) VALUES
                                (_?, CURRENT_TIMESTAMP, _?, _?, _?, _?)
    ''', [stage.idx, stage.apply_hash, stage.revert_hash,
          stage.apply_script, stage.revert_script])


# Updates an evolution row
def update_db(stage, dbConn):
    dbConn.execute('''
        UPDATE evolutions SET apply_hash = _?, revert_hash = _?,
                              apply_script = _?, revert_script = _?
                          WHERE id = _?
    ''', [stage.apply_hash, stage.revert_hash,
          stage.apply_script, stage.revert_script, stage.idx])


# Inserts or updates stages we were told to skip, return updated DB stages
def update_for_skips(dir_stages, db_stages, skip, dbConn):
    dir_len = len(dir_stages)
    db_len = len(db_stages)
    dir_i = 0
    db_i = 0
    for idx in sorted(list(skip)):
        # Find idx in dir and db
        while dir_i < dir_len:
            if dir_stages[dir_i].idx == idx:
                break
            dir_i += 1
        if dir_i == dir_len:
            raise Exception("Skip requested for missing file: %d.sql" % (idx))
        while db_i < db_len:
            if db_stages[db_i].idx == idx:
                # update
                dir_stages[dir_i].applied_at = db_stages[db_i].applied_at
                db_stages[db_i] = dir_stages[dir_i]
                update_db(db_stages[db_i], dbConn)
                break
            db_i += 1
            # If not found, will get added later under evolve() in sequence
    return db_stages


# 1. Go forward through DB rows to find first difference from files
# 2. Run downs from above down to and including that row, removing DB rows
# 3. Run files and insert DB rows starting from there
# 4. Return resulting updated DB stage objects
def evolve(dir_stages, db_stages, skip, prod_mode, dbConn):
    dir_len = len(dir_stages)
    db_len = len(db_stages)

    # Finds first differing stage, executes downs from DB up to that one,
    s = 0
    while s < min(dir_len, db_len):
        if dir_stages[s].apply_hash != db_stages[s].apply_hash:
            break
        s += 1

    # s is now index of first file that needs to be run
    # Run all DB downs in reverse through here
    downs_r = range(db_len-1, s-1, -1)
    if prod_mode and downs_r:
        raise Exception('In production mode but downs ' + str(downs_r)
                        +' needs running; aborting!')
    for i in downs_r:
        run_and_remove_downs(db_stages[i], dbConn)
    db_stages = db_stages[0:s]

    # Now run the ups
    for i in range(s, dir_len):
        run_and_add_ups(dir_stages[i], skip, dbConn)

    return db_stages[0:s] + dir_stages[s:]


def usage(invoked_name):
    print("usage: " + path.basename(invoked_name)
          + " <db_url> <db_user> <db_pass> <evolutions_dir> [--skip=<stages>] [--prod]\n"
          + "   --skip=<stages> = comma-separated indices to assume already run\n"
          + "   --prod          = abort if downs need to be run (for production)")
    return 1


def do_evolutions(ev_dir, skip, prod_mode, dbConn):

    # 1. Scan dir files in order, compute hashes (hashlib.sha1().hexdigest())
    dir_stages = check_stages(scan_dir_stages(ev_dir), ev_dir)
    if not dir_stages:
        raise Exception("No evolutions found in dir '" + ev_dir + "'")
    logger.info("Got %d stages from dir '%s'", len(dir_stages), ev_dir)
    logger.debug('\n\t%s', '\n\t'.join(map(str, dir_stages)))

    # 2. Scan DB files
    db_stages = check_stages(scan_db_stages(dbConn), 'db')
    logger.info("Got %d stages from DB '%s'", len(db_stages), dbConn.db_name)
    logger.debug('\n\t%s', '\n'.join(map(str, db_stages)))

    # 3. Handle skips
    db_stages = update_for_skips(dir_stages, db_stages, skip, dbConn)

    # 4. Evolve
    return evolve(dir_stages, db_stages, skip, prod_mode, dbConn)


def main(args):

    # Arg processing
    if len(args) < 5 or len(args) > 7:
        return usage(args[0])

    db_url, user, pw, ev_dir = args[1:5]

    skip_arg = '--skip='
    prod_arg = '--prod'
    skip = set()
    prod_mode = False
    for arg in args[5:]:
        if arg.startswith(skip_arg):
            skip = set(map(int, args[5][len(skip_arg):].split(',')))
            logger.info('Force skip: %s', (','.join(map(str, skip))))
        elif arg == prod_arg:
            prod_mode = True
        else:
            return usage(args[0])

    # Connect to DB and ensure evolutions table
    dbConn = connect_and_ensure(db_url, user, pw)

    ret = 0
    try:
        upd_stages = do_evolutions(ev_dir, skip, prod_mode, dbConn)
        logger.info('Completed with %d stages.', len(upd_stages))
    except Exception as e:
        logger.warning('Evolutions failed, check output and database;\n'
                       + '    exception was: %s', str(e))
        ret = 1
    finally:
        dbConn.conn.commit()
        dbConn.conn.close()

    return ret


if __name__ == "__main__":
    sys.exit(main(sys.argv[:]))
